File: elements-of-programming-interviews/2023/cpp/spreadsheet_encoding.cc
#include <cmath>
#include <string>

#include "test_framework/generic_test.h"
using std::string;

/*
A-Z = 1 to 26
AA = 27
ZZ = 702

- Not considering empty str
- This is a different kind of base 26; rightmost column is mod 26, next is mod
26*26.
- A to ZZ inclusive is 702 (26*26 + 26)
*/

#define ASCII_A 64  // so A = 1

int SSDecodeColIDWithExp(const string& col) {
  int total = 0;
  int exp = 0;
  for (auto it = col.rbegin(); it != col.rend(); it++, exp++) {
    total += ((*it - ASCII_A) * pow(26, exp));
  }
  return total;
}

int SSDecodeColID(const string& col) {
  int total = 0;
  for (auto it = col.begin(); it != col.end(); it++) {
    total = (total * 26) + *it - 'A' + 1;
  }
  return total;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"col"};
  return GenericTestMain(args, "spreadsheet_encoding.cc",
                         "spreadsheet_encoding.tsv", &SSDecodeColID,
                         DefaultComparator{}, param_names);
}

File: elements-of-programming-interviews/2023/python/spreadsheet_encoding.py
from test_framework import generic_test

"""
Input: string (could be large, but fits in mem)
Output: int (same constraints; represents string)

- "" invalid
- A = 1, Z = 26
- AA = 1*(26^0) + 1*(26*1)
- No 0, no negatives
"""
import string

DIGITS = {c: i+1 for i, c in enumerate(string.ascii_uppercase)}


def ss_decode_col_id(col: str) -> int:
    total = 0
    place = 1
    for c in col[::-1]:
        total += (DIGITS[c] * place)
        place *= 26
    return total


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('spreadsheet_encoding.py',
                                       'spreadsheet_encoding.tsv',
                                       ss_decode_col_id))
